airflow/utils/log/es_task_handler.py

```diff
@@ -255,7 +256,9 @@ def close(self):
 
         # Mark the end of file using end of log mark,
         # so we know where to stop while auto-tailing.
-        self.handler.stream.write(self.end_of_log_mark)
+        if self.write_stdout:
+            print()
+        self.handler.emit(logging.makeLogRecord({'msg': self.end_of_log_mark}))
 
```

Comment:
Doesn't this change where the end of log mark goes?

Reply:
yes, I sent email to them and suggested those changes are bad and copied that message to you. Please see email I sent to you and authors of 5528.

Reply:
sorry I meant that I sent the email to Andrii and the two authors of 5528.

Reply:
from 5528:
     When the end_of_log_mark is wrapped in a log record, the end_of_log_mark can no longer be 
     able to match the log line in _read:

        metadata['end_of_log'] = False if not logs \
            else logs[-1].message == self.end_of_log_mark.strip()
     It leads to the UI keeps calling backend and generates lots of load to ES.
By removing the log_id from the end-of-log mark, it would make it worse as the ui would continue to try to find the end-of-log mark and it won't ever find it as it searches the end-of-log mark by log_id.

I am not sure what  the sentence mean by "When the end_of_log_mark is wrapped in a log record". I also observed that the end-of-log mark might end up within the same line of other log lines and it would prevent us from find the end-of-log mark in those cases. To address that, I always add an obnoxious print right in front of the end-of-log mark line, to ensure the "end-of-log" mark is always in a separate line when printing to console. This is import for filebeat/logstash on kubernetes to pick up the end-of-log mark log line in a separate document.

Reply:
It means the last line would be something like `[2020-01-16 07:58:32,712] {es_task_handler.py:XXX} INFO [end_of_log_mark]` and thus made the reader unable to understand it.

I'm a bit lost how did this removed the log id from the end_of_log_mark. Isn't the log_id we constructed in this file only for log fetching? My understanding is that the log_id is determined when we upload the log, e.g. when we pipe stdout to logstash or when we upload file through filebeat to logstash.

Maybe I was understanding this wrong and there is indeed a bug. In that case I would agree on spliting this change into two PRs for sanity purpose.

Reply:
Kevin, the removal of end-of-log mark is already handled by Andrii's initial changes adding the stdout support: https://github.com/apache/airflow/commit/0da976a0e1e28e2c0cd274d7384cf2976db6deec#diff-485751b55125e8a90050d22f69e8467c

        # end_of_log_mark may contain characters like '\n' which is needed to
        # have the log uploaded but will not be stored in elasticsearch.
        metadata['end_of_log'] = False if not logs \
            else logs[-1].message == self.end_of_log_mark.strip()

then
        # If we hit the end of the log, remove the actual end_of_log message
        # to prevent it from showing in the UI.
        i = len(logs) if not metadata['end_of_log'] else len(logs) - 1
        message = '\n'.join([log.message for log in logs[0:i]])

Please see my test case test_close_with_log_id that exercises this logic in the tests now.
Can you please check if this is clear to you now?

Reply:
I did observe that occasionally the end-of-log mark is not on its separate line in the console, then it can confuse elasticsearch. So I always add a print() in this PR right before the end-of-log print out and this is just a reliability fix given we need the end-of-log mark in a separate log line for this to work.

Reply:
with regarding to request to split this change into two PRs, can you please check my test case test_close_with_log_id. I need to use the separate index parameter  to safely test the log_id logic correctly as I showed above, and ensure the end-of-log mark is removed correctly and the # of messages are expected etc. It is needed in this PR to improve testability of the code. Can you please check if this is reasonable?

Reply:
```
 else logs[-1].message == self.end_of_log_mark.strip()
```
I think this will not work if the `end_of_log_mark` is wrapped as a log record, as the `message` will be the format that Kevin mentioned and it also depends on the `log formatter`.

Reply:
this works as long as you do not put white spaces into your end_of_log_mark. I think you would just ask for troubles by putting white space characters into the end-of-log mark.

Reply:
sry I was get lost in my inbox, yes it is very clear to me now. Thank you for filling me in with details!

Reply:
Hi @larryzhu2018 for your description: 
![image](https://user-images.githubusercontent.com/8662365/72946940-52353080-3d35-11ea-9c3b-47996b94f47f.png)

It looks like you have `write_stdout` as `true`, in fact, my pr: https://github.com/apache/airflow/pull/7199 fixed the always `true` for `write_stdout`.  could you please test the case when `write_stdout` is `False`.

Our set up is:
`write_stdout` is `False`
`json` is `False`

Reply:
> this works as long as you do not put white spaces into your end_of_log_mark. I think you would just ask for troubles by putting white space characters into the end-of-log mark.

I am not sure why using white space will cause troubles.

Reply:
> Hi @larryzhu2018 for your description:
> ![image](https://user-images.githubusercontent.com/8662365/72946940-52353080-3d35-11ea-9c3b-47996b94f47f.png)
> 
> It looks like you have `write_stdout` as `true`, in fact, my pr: #7199 fixed the always `true` for `write_stdout`. could you please test the case when `write_stdout` is `False`.
[larryzhu2018]: the test case I have in the PR, test_close_with_log_id, is setting write_stdout to False so please check.
> 
> Our set up is:
> `write_stdout` is `False`
> `json` is `False`
> `END_OF_LOG_MARK = u'\u0004\n'`
> 
> This is what we see in the Kibana
> ![image](https://user-images.githubusercontent.com/8662365/72947738-d8eb0d00-3d37-11ea-8eac-d341a2c66163.png)
> 
> I remember if the `end_of_log` is wrapped in a `logging.makeLogRecord` it will start with
> 
> ![image](https://user-images.githubusercontent.com/8662365/72947783-033cca80-3d38-11ea-9fc2-a31429464412.png)

Reply:
> > this works as long as you do not put white spaces into your end_of_log_mark. I think you would just ask for troubles by putting white space characters into the end-of-log mark.
> 
> I am not sure why using white space will cause troubles.
see the .strip() call as I quoted earlier. This basically means if you have empty spaces in your end-of-log mark, it won't work and it will confuse the users.

 else logs[-1].message == self.end_of_log_mark.strip()

Does this make sense now?

But please do not remove .strip() call to "fix" this. It is hard to guarantee the event pipelines preserve whitespaces so it is best to not use whitespace in your end-of-log mark.

Reply:
> ```
>  else logs[-1].message == self.end_of_log_mark.strip()
> ```
> 
> I think this will not work if the `end_of_log_mark` is wrapped as a log record, as the `message` will be the format that Kevin mentioned and it also depends on the `log formatter`.
I did not change any production code but did add a test case to show how this works. 

in short, you will need to deploy an ingest log processor, and the ones we have look like this:
"description" : "cluster json log Pipeline",
"processors" : [
  {
    "rename" : {
      "field" : "message",
      "target_field" : "raw_message"
    }
  },
  {
    "json" : {
      "field" : "raw_message",
      "add_to_root" : false,
      "target_field" : "json_target"
    }
  },
  {
    "grok" : {
      "field" : "json_target.message",
      "patterns" : [
        "Job %{DATA:job_id}: Subtask %{DATA} %{GREEDYDATA:json_msg}",
        "%{GREEDYDATA}"
      ]
    }
  },
  {
    "json" : {
      "field" : "json_msg",
      "add_to_root" : true,
      "if" : "ctx.job_id != null"
    }
  },
  {
    "json" : {
      "field" : "raw_message",
      "add_to_root" : true,
      "if" : "ctx.job_id == null"
    }
  },
  {
    "remove" : {
      "field" : "json_msg",
      "ignore_missing" : true
    }
  },
  {
    "remove" : {
      "field" : "json_target"
    }
  },
  {
    "set" : {
      "field" : "event.kind",
      "value" : "tasks",
      "if" : "ctx.message != null"
    }
  },
  {
    "set" : {
      "field" : "event.dataset",
      "value" : "airflow",
      "if" : "ctx.dag_id != null && ctx.task_id != null"
    }
  },
  {
    "set" : {
      "field" : "log_id",
      "value" : "{{dag_id}}-{{task_id}}-{{execution_date}}-{{try_number}}",
      "if" : "ctx.event?.dataset == 'airflow'"
    }
  },
  {
    "set" : {
      "field" : "offset",
      "value" : "{{log.offset}}",
      "if" : "ctx.event?.dataset == 'airflow'"
    }
  }
],
"on_failure" : [
  {
    "set" : {
      "field" : "error.message",
      "value" : "{{ _ingest.on_failure_message }}"
    }
  }
]